pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -140,8 +154,51 @@ func (gc *GarbageCollector) Run(workers int, stopCh <-chan struct{}) {
 	<-stopCh
 }
 
-func (gc *GarbageCollector) HasSynced() bool {
-	return gc.dependencyGraphBuilder.HasSynced()
+// ResettableRESTMapper is a RESTMapper which is capable of resetting itself
+// from discovery.
+type ResettableRESTMapper interface {
+	meta.RESTMapper
+	Reset()
+}
+
+// Sync periodically resyncs the garbage collector monitors with resources
+// returned found via the discoveryClient. Sync blocks, continuing to sync until
+// a message is received on stopCh.
+//
+// The discoveryClient should be the same client which underlies restMapper.
+func (gc *GarbageCollector) Sync(restMapper ResettableRESTMapper, discoveryClient discovery.DiscoveryInterface, period time.Duration, stopCh <-chan struct{}) {
+	t := time.NewTicker(period)
```

Comment:
I'd recommend using `wait.Until` to periodically run until the stop channel is closed, instead of doing your own ticker/for loop.